[Woo POS] [Cash & Receipts] Better support for large font sizes

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14792

## Description
This PR improves support for large font sizes and makes the POS cash collection flow more accessible, mostly by being sure that buttons are not hidden when using large accessibility sizes.

## General feedback
I've piggy-bagged on how we do it in the rest of POS (eg: mostly relied on dynamicTypeSize checks), I'm not fully aware of what options we have here as I haven't done much work around a11y, happy to iterate as needed while I check WWDC's for tips 😅 

## Changes

* Problem: The `Cash Payment` button is outside of the screen -> Solution: We remove the VStack spacing when isAccessibilitySize 280506ce0e2d0e27918f232302afde71cc6f3b2b

| Before | After |
|--------|--------|
| ![_large_font_size_0](https://github.com/user-attachments/assets/3e045e01-5f04-4cf9-b6f6-02f53e40f254) | ![__large_font_size_cash_payment_button](https://github.com/user-attachments/assets/ab21913d-7549-4f30-ad93-db2ee9ed8f4b) | 

* Problem: The `Mark payment as complete` button is hidden entirely when the keyboard is present -> Solution: Made 2 changes: 
  * Validate/mark complete on `.onSubmit()` when the keyboard is open
  * Update the layout to use the horizontal space better and remove space so we can fit all components.

| Before | After |
|--------|--------|
| ![_large_font_size_cash_payment_show_keyboard_0](https://github.com/user-attachments/assets/83d41569-6a66-4d9b-8de8-4e5be123f6ac) | ![simulator_screenshot_5E19DED3-2C3F-46DF-9F58-309A5B55A4D0](https://github.com/user-attachments/assets/fd40731c-66f0-4a07-bea2-3c8a6af6d2a4) | 


## Testing information
- Enable `.acceptCashForPointOfSale`
- Set the device or simulator to use large font sizes (I've used `cmd/option/+` until couldn't be bigger)
- Go through a POS and confirm that you can process a cash transaction and send a receipt and no buttons are hidden due to view composition.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
remove vstack spacing when a11y size

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/235">21.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
validate-submit from keyboard on value submission

In case the submission button gets hidden due to font sizes, we can also submit the values directly from the keyboard for better UX

iamgabrielma committed:
change header distribution when large font size

iamgabrielma committed:
remove vstack spacing when needed. apply scale effect

The FormattableAmountTextField is currently way too  big even for large fonts, in detriment of being able to add more components onscreen when needed. We keep the defauls size, but reduce it slightly in larger a11y sizes

iamgabrielma committed:
collapse func to avoid duplication

iamgabrielma committed:
introduce conditionalPadding

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -120,6 +106,55 @@ struct PointOfSaleCollectCashView: View {
 private extension PointOfSaleCollectCashView {
+    @ViewBuilder
+    var navigationHeader: some View {
+        if dynamicTypeSize.isAccessibilitySize {
```
I tried to collapse this one into smaller subviews but ended causing more problems than anything else as they ended being too simple views, or need to add many conditionals so the design would stick in non-accessibility mode, so I just left the duplication for simplicity.

staskus replied:
We have `DynamicHStack` for this exact need but the opposite case 😀 It switches from HStack to VStack, and you want to so switch from VStack to HStack. But feel free to create equivalent `DynamicVStack` and use it in this case. 


```
 HStack(alignment: .top) {
                Image(systemName: "chevron.backward")
                    .font(.posBodyEmphasized, maximumContentSizeCategory: .accessibilityLarge)
                DynamicVStack(alignment: .leading) {
                    Text(Localization.backNavigationTitle)
                        .font(.posTitleEmphasized)
                        .accessibilityAddTraits(.isHeader)

                    Text(formattedOrderTotal)
                        .font(.posBodyRegular)
                }
                .padding(.top, -Constants.navigationButtonSpacing)
            }
            .foregroundColor(navigationForegroundColor)
```

iamgabrielma replied:
Oh, that's nice! I've duplicated it for VStack and works great https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/6f9f3fa9830de6d2c3f79892f9cc6ad4b642072c

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -147,6 +182,10 @@ private extension PointOfSaleCollectCashView {
         static let buttonCornerRadius: CGFloat = 8
     }
 
+    private func conditionalPadding(_ padding: CGFloat) -> CGFloat {
+        dynamicTypeSize.isAccessibilitySize ? 0 : padding
+    }
```
By setting the value to `0` it's a simple approach I found mostly works without modifying what we already have (this seems to fit other places in POS I've checked). 

Initially I tried to leverage using `@ScaledMetric` and multiply this by the `padding` we pass in when `!isAccessibilitySize`  rather than just returning either `0` or the `padding` value, something like:
```
    private func conditionalPadding(_ padding: CGFloat) -> CGFloat {
        debugPrint("scale")
        debugPrint("\(dynamicTypeSize) is accessibility size? \(dynamicTypeSize.isAccessibilitySize)")
        /**
         xxxlarge false
         accessibility1 true scale: 1.5
         accessibility2 true scale: 2
         accessibility3 true scale: 2
         accessibility4 true scale: 2.5
         */
        // Won't work right away because not all values are accessibility sizes, and some use the same scaled metric
        return dynamicTypeSize.isAccessibilitySize ? (padding * scale) : padding
```

However, as we have to modify the default font several times before it moves from `xxlarge` > `xxxlarge` to accessibility values (`accessibility1`, `accessibility2`), etc, we would need to adapt all sizes for a smooth transition for each as would also affect smaller sizes.

Perhaps we can do something like the `POSModalSizing` across some POS components? 🤔

staskus replied:
> Perhaps we can do something like the POSModalSizing across some POS components? 🤔

I don't have an opinion on this. 

I think your chosen approach is good. We want to have more space for what matters - text and buttons, while limiting less important things - spacing.

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -24,33 +25,27 @@ struct PointOfSaleCollectCashView: View {
             }
 
             FormattableAmountTextField(viewModel: textFieldViewModel, style: .pos)
+                .scaleEffect(dynamicTypeSize.isAccessibilitySize ? 0.7 : 1.0)
```
This modifier can also be used to restrict some views from growing past a defined size. Some views may be large enough by themselves. 
```suggestion
                .dynamicTypeSize(...DynamicTypeSize.accessibility3)
```

iamgabrielma replied:
💯  I didn't know that `...upperbound` syntax. I've updated it to use this instead https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/3d3c077f9d2444eb33b249a9f31270edffacd8b2

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -70,16 +65,7 @@ struct PointOfSaleCollectCashView: View {
-                        errorMessage = Localization.failedToCollectCashPayment
-                    }
-                    isLoading = false
+                    await submitCashAmount()
```
Right now the sizes of some elements become smaller when we increase the text size. If they don't fit anymore, we can just limit their size. In the case of button and text field.

https://github.com/user-attachments/assets/be7ee068-679f-43e9-8972-66f3587ae247

iamgabrielma replied:
I've played with .dynamicTypeSize a bit here https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/3d3c077f9d2444eb33b249a9f31270edffacd8b2
and here https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/5d5af43f52a98e643794f39e25865875b34c39b9 . By limiting the upper bounds to `accessibility3` and `accessibility1` works nicely for iPads and mini 👍

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -70,16 +65,7 @@ struct PointOfSaleCollectCashView: View {
-                        errorMessage = Localization.failedToCollectCashPayment
-                    }
-                    isLoading = false
+                    await submitCashAmount()
```
There's also a case where the keyboard starts hiding the button. We could also check with iPad Mini, the keyboard may be hiding the whole button in some cases.

In `UIKit` days, we used to create reusable functionality by tracking the keyboard size and shrinking the view in the process. Maye there's a recommendation of a similar `SwiftUI` modifier.  Usually in these scenarios scroll views were used. In this case, we have a a header, text field, and a button. We should be able to fit the elements without using a scroll view.

![IMG_0157](https://github.com/user-attachments/assets/3ed62778-def4-4c95-a015-30fcbcd61447)

iamgabrielma replied:
> There's also a case where the keyboard starts hiding the button. We could also check with iPad Mini, the keyboard may be hiding the whole button in some cases.

That was the case indeed. In the case of iPad mini I'm setting the padding to 0 for all cases so it fits nicely. [The device check](https://github.com/woocommerce/woocommerce-ios/blob/6b52c7c3aa4c0f1d0f9df4056125b8462e5a6448/WooCommerce/Classes/POS/Presentation/TotalsView.swift#L398) feels a bit like a hack thought.

![_ipad_mini](https://github.com/user-attachments/assets/2b200fa6-bb7b-4899-af4c-8faa99a58fab)

staskus replied:
> [The device check](https://github.com/woocommerce/woocommerce-ios/blob/6b52c7c3aa4c0f1d0f9df4056125b8462e5a6448/WooCommerce/Classes/POS/Presentation/TotalsView.swift#L398) feels a bit like a hack thought.

Yes, it does. We should try to avoid it. Maybe we could control spacing not with VStack but with Spacers who have maxHeight and they would shrink if elements don't fit 🤔 . Of course need to play around.

staskus commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -34,7 +34,7 @@ struct TotalsView: View {
                         .renderedIf(cardReaderViewLayout.topPadding == nil)
 
-                    VStack(alignment: .center, spacing: Constants.verticalSpacing) {
+                    VStack(alignment: .center, spacing: dynamicTypeSize.isAccessibilitySize ? 0 : Constants.verticalSpacing) {
```
For larger non-accessibility sizes, the button may not fit.

This is the only blocking comment. The other ones are optional improvements.

![IMG_0156](https://github.com/user-attachments/assets/c823fca6-a295-4c87-8bc6-85e69d3b87d6)

iamgabrielma replied:
That's a very good catch, I've updated the dynamic spacing so it's more nuanced and takes into account non-accessibility sizes as well: https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/16cff5f4ec75faa038d36921e338c02943991f81 and it shows correctly for x+ sizes before moving to accessibility sizes:

| xlarge | xxlarge | xxxlarge |
|--------|--------|--------|
| ![_xlarge](https://github.com/user-attachments/assets/9fe34456-9be9-477a-8d28-4e14864ff15f) | ![_xxlarge](https://github.com/user-attachments/assets/373b9748-8fdc-4123-a96c-67282d8e0cd0) | ![_xxxlarge](https://github.com/user-attachments/assets/1cfffed9-99ad-46d8-8edb-78dc94271f23) | 

I'll check the other items now and test on iPad Mini as well to fine-tune this a bit better where possible 👍

staskus reviewed (commented):
Thanks for the work!

Made a few suggestions, one related to Totals View I think should be done now. 

I didn't test Email Receipt view. Let me know if I should.

staskus reviewed (approved):
I'm approving so not to block the merge once you make improvement. Feel free to ask for another review if needed.

iamgabrielma committed:
Account for non-accessibility dynamic type sizes

The previous implementation worked for .accessibility 1-to-5, but didn’t quite work for xxLarge+ as cash button was still pushed outside the view partially

iamgabrielma committed:
fine-tune vertical spacing for iPad mini

iamgabrielma committed:
Use DynamicVStack to avoid duplication

iamgabrielma committed:
Restrict text field size upper bound

iamgabrielma committed:
Adjust textfield and button upper bounds for smaller devices

We can go up to accessibility3 in iPad11, but for iPad mini this is too big, setting the upper bound on accessibility1 feels adequate for both

iamgabrielma committed:
Merge branch 'trunk' into task/14792-pos-cash-larger-font-sizes

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift

iamgabrielma committed:
Adapt receipts view

iamgabrielma commented:
Thanks for the review Povilas!

> I didn't test Email Receipt view. Let me know if I should.

Since it has less vertical elements this was mostly fine, but I've given it a pass on https://github.com/woocommerce/woocommerce-ios/pull/14935/commits/6b52c7c3aa4c0f1d0f9df4056125b8462e5a6448 and modified it nonetheless, so we're sure it works for smaller devices. Switching between some of the sizes jumps a bit, but buttons are never blocked:

https://github.com/user-attachments/assets/5a3e3e24-89e4-4c65-8ab7-69965f6a5451

I'll leave this here till tomorrow before merging in case you want to take another look, but I don't think is necessary. Feel free depending on priorities 👍

staskus commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -34,7 +34,7 @@ struct TotalsView: View {
                         .renderedIf(cardReaderViewLayout.topPadding == nil)
 
-                    VStack(alignment: .center, spacing: Constants.verticalSpacing) {
+                    VStack(alignment: .center, spacing: dynamicVerticalSpacing(for: dynamicTypeSize)) {
```
As mentioned in another comment, I would experiment setting the spacing to 0, and instead playing with Spacer() elements. There should be a way not to use concrete device checking.

iamgabrielma replied:
You were totally right here, by setting the spacing to 0 we seem to resolve the issue and we don't even need to play with additional or existing Spacers. The button its slightly cut in the higher font size with the smaller device but is perfectly usable. I'll leave it like this to not overcomplicate it.

iPad 11

https://github.com/user-attachments/assets/b5ddff54-8361-4442-b17f-132918a901e9

iPad mini

https://github.com/user-attachments/assets/0fa28429-1efe-42c8-9d1f-01793ec971d2

Upon exploring I've also discovered that since iOS16 we have [ViewThatFits](https://developer.apple.com/documentation/swiftui/viewthatfits), which would use different child views based on the actual size. It's a bit late for this specific case, but looks like a great component to use when we have the chance to avoid too much complexity around dynamic sizes and to even iterate on design feedback around what should be included and how for smaller screens/bigger components.

iamgabrielma committed:
remove iPadMini check and set vertical spacing to 0
